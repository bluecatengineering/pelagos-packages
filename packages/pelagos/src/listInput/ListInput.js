import {useCallback, useRef, useState} from 'react';
import PropTypes from 'prop-types';
import {t} from '@bluecateng/l10n.macro';

import useRandomId from '../hooks/useRandomId';
import LabelLine from '../components/LabelLine';
import ComboBox from '../components/ComboBox';
import FieldHelper from '../formFields/FieldHelper';

import ListEntries from './ListEntries';

import './ListInput.less';

/** An input box for a list of values. */
const ListInput = ({
	id,
	className,
	label,
	required,
	disabled,
	emptyText,
	list,
	helperText,
	error,
	reorderable,
	column,
	getSuggestions,
	getSuggestionText,
	getSuggestionValue,
	getHighlightKey,
	parseInput,
	validateSuggestion,
	renderSuggestion,
	getItemKey,
	getItemName,
	renderItem,
	onListChange,
	onTextChange,
	onErrorChange,
	...props
}) => {
	id = useRandomId(id);

	const liveRef = useRef(null);

	const [text, setText] = useState('');
	const [highlightKey, setHighlightKey] = useState(null);

	const handleGetSuggestions = useCallback(
		(text) => {
			const {suggestions, error} = getSuggestions(text, list);
			if (error) {
				onErrorChange(error);
				return [];
			}

			return suggestions;
		},
		[list, getSuggestions, onErrorChange]
	);

	const handleChange = useCallback(
		(suggestion) => {
			const id = getHighlightKey(suggestion);
			if (id) {
				setText('');
				onTextChange(false);
				setHighlightKey(id);
			} else if (parseInput) {
				let name, item;
				if (getSuggestionValue) {
					item = getSuggestionValue(suggestion);
					name = getItemName(item);
				} else {
					item = name = getSuggestionText(suggestion);
				}
				liveRef.current.textContent = t`${name} added`;
				setText('');
				onTextChange(false);
				onListChange([item, ...list]);
			} else {
				const item = getSuggestionValue ? getSuggestionValue(suggestion) : suggestion;
				const name = getSuggestionValue ? getItemName(item) : getSuggestionText(suggestion);
				const error = validateSuggestion && validateSuggestion(suggestion);
				if (error) {
					setText(name);
					onTextChange(!!name);
					onErrorChange(error);
				} else {
					liveRef.current.textContent = t`${name} added`;
					setText('');
					onTextChange(false);
					onListChange([item, ...list]);
				}
			}
		},
		[
			getHighlightKey,
			parseInput,
			onTextChange,
			onListChange,
			getSuggestionText,
			getSuggestionValue,
			getItemName,
			list,
			validateSuggestion,
			onErrorChange,
		]
	);

	const handleEnter = useCallback(
		(text) => {
			if (text === '/clear') {
				setText('');
				onTextChange(false);
				onListChange([]);
			} else if (text && parseInput) {
				const {entries, error} = parseInput(text, list);
				if (error) {
					onErrorChange(error);
				} else {
					const name = entries.map(getItemName).join(', ');
					liveRef.current.textContent = t`${name} added`;
					setText('');
					onTextChange(false);
					onListChange([...entries, ...list]);
					onErrorChange(null);
				}
			}
		},
		[list, parseInput, getItemName, onTextChange, onListChange, onErrorChange]
	);

	const handleTextChange = useCallback(
		(text) => {
			if (error) {
				onErrorChange(null);
			}
			setText(text);
			onTextChange(!!text);
		},
		[error, onTextChange, onErrorChange]
	);

	const handleRemoveClick = useCallback(
		(item, index) => onListChange([...list.slice(0, index), ...list.slice(index + 1)]),
		[list, onListChange]
	);

	const handleHighlightClear = useCallback(() => setHighlightKey(null), []);

	const empty = !list?.length;
	const labelId = `${id}-label`;
	const helperId = `${id}-helper`;
	return (
		<div className={`ListInput${className ? ` ${className}` : ''}`}>
			<div className="sr-only" aria-live="assertive" ref={liveRef} />
			<LabelLine id={labelId} htmlFor={id} text={label} required={required} error={!!error} />
			<ComboBox
				{...props}
				id={id}
				autoSelect={!parseInput}
				text={text}
				disabled={disabled}
				error={!!error}
				getSuggestions={handleGetSuggestions}
				renderSuggestion={renderSuggestion}
				aria-describedby={helperId}
				aria-required={required}
				onChange={handleChange}
				onEnter={handleEnter}
				onTextChange={handleTextChange}
			/>
			<FieldHelper id={helperId} text={helperText} error={error} />
			{empty ? (
				<div className="ListInput__empty" id={id + '-empty'} aria-live="polite">
					{emptyText}
				</div>
			) : (
				<ListEntries
					id={id + '-grid'}
					className="ListInput__list"
					highlightKey={highlightKey}
					list={list}
					disabled={disabled}
					reorderable={reorderable}
					column={column}
					getItemKey={getItemKey}
					getItemName={getItemName}
					renderItem={renderItem}
					onReorder={onListChange}
					onRemoveClick={handleRemoveClick}
					onHighlightClear={handleHighlightClear}
				/>
			)}
		</div>
	);
};

ListInput.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The label text. */
	label: PropTypes.string,
	/** Whether the field is required. */
	required: PropTypes.bool,
	/** Whether the field is disabled. */
	disabled: PropTypes.bool,
	/** The placeholder text in the input box. */
	placeholder: PropTypes.string,
	/** The text below input box when list is empty. */
	emptyText: PropTypes.string,
	/** The entries for the list. */
	list: PropTypes.array,
	/** Additional information for the field. */
	helperText: PropTypes.string,
	/** The error text. */
	error: PropTypes.string,
	/** Whether the list should be reorderable. */
	reorderable: PropTypes.oneOf([true, false, 'numbers']),
	/** Whether the list should be displayed as columns. Set to true if reorderable. */
	column: PropTypes.bool,
	/** Function invoked to provide a list of suggestions. */
	getSuggestions: PropTypes.func,
	/** Function invoked to render the suggestion text for the returned items. (see above) */
	getSuggestionText: PropTypes.func,
	/** Function invoked to get the value from the suggestion to add to the list. (see above) */
	getSuggestionValue: PropTypes.func,
	/** Function invoked to get the key of the suggestion to highlight. */
	getHighlightKey: PropTypes.func,
	/** Function invoked to parse the input text. */
	parseInput: PropTypes.func,
	/** Function invoked to validate the suggestion. */
	validateSuggestion: PropTypes.func,
	/** Function invoked to render the suggestions. */
	renderSuggestion: PropTypes.func,
	/** Function invoked to get each item's key. */
	getItemKey: PropTypes.func.isRequired,
	/** Function invoked to get each item's name. */
	getItemName: PropTypes.func.isRequired,
	/** Function invoked to render each list item. */
	renderItem: PropTypes.func,
	/** Function invoked when the list changes. */
	onListChange: PropTypes.func.isRequired,
	/** Function invoked when the text input changes. */
	onTextChange: PropTypes.func.isRequired,
	/** Function invoked when error status changes. */
	onErrorChange: PropTypes.func,
};

export default ListInput;
