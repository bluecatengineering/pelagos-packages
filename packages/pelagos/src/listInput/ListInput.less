@import '../../less/fonts';
@import '../../less/spacing';

@layer pelagos {
	.ListInput {
		margin-top: @sp-12;
	}

	.ListInput__empty {
		@helper-text-02();
		color: var(--text-helper);
		white-space: pre-line;
		height: 145px;
		text-align: center;
	}

	.ListInput__list {
		height: 145px;
		overflow-y: auto;
	}
}
