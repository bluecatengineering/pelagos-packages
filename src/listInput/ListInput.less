@import '../../less/colors';
@import '../../less/fonts';
@import '../../less/spacing';

.ListInput {
	margin-top: @sp-12;
}

.ListInput__empty {
	@font-16-400();
	color: @font-color-3;
	white-space: pre-line;
	height: 145px;
	text-align: center;
}
