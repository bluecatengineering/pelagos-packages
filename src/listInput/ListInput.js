import {useCallback, useState} from 'react';
import PropTypes from 'prop-types';

import LabelLine from '../components/LabelLine';
import ComboBox from '../components/ComboBox';
import FieldError from '../formFields/FieldError';
import setLiveText from '../functions/setLiveText';
import __ from '../strings';

import ListEntries from './ListEntries';

import './ListInput.less';

/** An input box for a list of values. */
const ListInput = ({
	id,
	label,
	optional,
	notice,
	emptyText,
	list,
	error,
	column,
	getSuggestions,
	getSuggestionText,
	getHighlightKey,
	parseInput,
	validateSuggestion,
	renderSuggestion,
	getItemKey,
	getItemName,
	renderItem,
	onListChange,
	onTextChange,
	onErrorChange,
	...props
}) => {
	const [text, setText] = useState('');
	const [highlightKey, setHighlightKey] = useState(null);

	const handleGetSuggestions = useCallback(
		(text) => {
			const {suggestions, error} = getSuggestions(text, list);
			if (error) {
				onErrorChange(error);
				return [];
			}

			return suggestions;
		},
		[list, getSuggestions, onErrorChange]
	);

	const handleChange = useCallback(
		(suggestion) => {
			const id = getHighlightKey(suggestion);
			if (id) {
				setText('');
				onTextChange(false);
				setHighlightKey(id);
			} else {
				const suggestionText = getSuggestionText(suggestion);
				if (parseInput) {
					setLiveText(__('OBJECT_ADDED', {name: suggestionText}));
					setText('');
					onTextChange(false);
					onListChange([suggestionText, ...list]);
				} else {
					const error = validateSuggestion && validateSuggestion(suggestion);
					if (error) {
						setText(suggestionText);
						onTextChange(!!suggestionText);
						onErrorChange(error);
					} else {
						setLiveText(__('OBJECT_ADDED', {name: suggestionText}));
						setText('');
						onTextChange(false);
						onListChange([suggestion, ...list]);
					}
				}
			}
		},
		[
			getHighlightKey,
			parseInput,
			onTextChange,
			onListChange,
			getSuggestionText,
			list,
			validateSuggestion,
			onErrorChange,
		]
	);

	const handleEnter = useCallback(() => {
		if (text === '/clear') {
			setText('');
			onTextChange(false);
			onListChange([]);
		} else if (text && parseInput) {
			const {entries, error} = parseInput(text, list);
			if (error) {
				onErrorChange(error);
			} else {
				setLiveText(__('OBJECT_ADDED', {name: entries.join(', ')}));
				setText('');
				onTextChange(false);
				onListChange([...entries, ...list]);
				onErrorChange(null);
			}
		}
	}, [list, text, parseInput, onTextChange, onListChange, onErrorChange]);

	const handleTextChange = useCallback(
		(text) => {
			if (error) {
				onErrorChange(null);
			}
			setText(text);
			onTextChange(!!text);
		},
		[error, onTextChange, onErrorChange]
	);

	const handleRemoveClick = useCallback((item) => onListChange(list.filter((old) => old !== item)), [
		list,
		onListChange,
	]);

	const handleHighlightClear = useCallback(() => setHighlightKey(null), []);

	const empty = !list?.length;
	return (
		<div className="ListInput">
			<LabelLine htmlFor={id} text={label} optional={optional && empty} notice={notice} />
			<ComboBox
				{...props}
				id={id}
				autoSelect={!parseInput}
				text={text}
				error={!!error}
				getSuggestions={handleGetSuggestions}
				renderSuggestion={renderSuggestion}
				onChange={handleChange}
				onEnter={handleEnter}
				onTextChange={handleTextChange}
			/>
			<FieldError id={id + '-error'} text={error} />
			{empty ? (
				<div className="ListInput__empty" id={id + '-empty'}>
					{emptyText}
				</div>
			) : (
				<ListEntries
					id={id + '-grid'}
					highlightKey={highlightKey}
					list={list}
					column={column}
					getItemKey={getItemKey}
					getItemName={getItemName}
					renderItem={renderItem}
					onRemoveClick={handleRemoveClick}
					onHighlightClear={handleHighlightClear}
				/>
			)}
		</div>
	);
};

ListInput.propTypes = {
	/** The component id. */
	id: PropTypes.string.isRequired,
	/** The label text. */
	label: PropTypes.string,
	/** Whether the field is optional. */
	optional: PropTypes.bool,
	/** The notice/warning text. */
	notice: PropTypes.string,
	/** The placeholder text in the input box. */
	placeholder: PropTypes.string,
	/** The text below input box when list is empty. */
	emptyText: PropTypes.string,
	/** The entries for the list. */
	list: PropTypes.array,
	/** The error text. */
	error: PropTypes.string,
	/** Whether the list should be displayed as columns. */
	column: PropTypes.bool,
	/** Function invoked to provide a list of suggestions. */
	getSuggestions: PropTypes.func,
	/** Function invoked to render the suggestion text for the returned items. (see above) */
	getSuggestionText: PropTypes.func,
	/** Function invoked to get the key of the suggestion to highlight. */
	getHighlightKey: PropTypes.func,
	/** Function invoked to parse the input text. */
	parseInput: PropTypes.func,
	/** Function invoked to validate the suggestion. */
	validateSuggestion: PropTypes.func,
	/** Function invoked to render the suggestions. */
	renderSuggestion: PropTypes.func,
	/** Function invoked to get each item's key. */
	getItemKey: PropTypes.func,
	/** Function invoked to get each item's name. */
	getItemName: PropTypes.func,
	/** Function invoked to render each list item. */
	renderItem: PropTypes.func,
	/** Function invoked when the list changes. */
	onListChange: PropTypes.func,
	/** Function invoked when the text input changes. */
	onTextChange: PropTypes.func,
	/** Function invoked when error status changes. */
	onErrorChange: PropTypes.func,
};

export default ListInput;
