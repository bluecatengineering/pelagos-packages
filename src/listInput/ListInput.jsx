import React, {useCallback, useState} from 'react';
import PropTypes from 'prop-types';

import Label from '../components/Label';
import ComboBox from '../components/ComboBox';
import FieldError from '../formFields/FieldError';

import ListEntries from './ListEntries';

import './ListInput.less';

const ListInput = ({
	id,
	label,
	notice,
	emptyText,
	list,
	error,
	column,
	getSuggestions,
	getSuggestionText,
	getHighlightKey,
	parseInput,
	validateSuggestion,
	renderSuggestion,
	getItemKey,
	renderItem,
	onListChange,
	onTextChange,
	onErrorChange,
	...props
}) => {
	const [text, setText] = useState('');
	const [highlightKey, setHighlightKey] = useState(null);

	const handleGetSuggestions = useCallback(
		text => {
			const {suggestions, error} = getSuggestions(text, list);
			if (error) {
				onErrorChange(error);
				return [];
			}

			return suggestions;
		},
		[list, getSuggestions, onErrorChange]
	);

	const handleChange = useCallback(
		suggestion => {
			const id = getHighlightKey(suggestion);
			if (id) {
				setText('');
				onTextChange(false);
				setHighlightKey(id);
			} else if (parseInput) {
				setText('');
				onTextChange(false);
				onListChange([getSuggestionText(suggestion), ...list]);
			} else {
				const error = validateSuggestion && validateSuggestion(suggestion);
				if (error) {
					const suggestionText = getSuggestionText(suggestion);
					setText(suggestionText);
					onTextChange(!!suggestionText);
					onErrorChange(error);
				} else {
					setText('');
					onTextChange(false);
					onListChange([suggestion, ...list]);
				}
			}
		},
		[
			getHighlightKey,
			parseInput,
			onTextChange,
			onListChange,
			getSuggestionText,
			list,
			validateSuggestion,
			onErrorChange,
		]
	);

	const handleEnter = useCallback(() => {
		if (text === '/clear') {
			setText('');
			onTextChange(false);
			onListChange([]);
		} else if (text && parseInput) {
			const {entries, error} = parseInput(text, list);
			if (error) {
				onErrorChange(error);
			} else {
				setText('');
				onTextChange(false);
				onListChange([...entries, ...list]);
			}
		}
	}, [list, text, parseInput, onTextChange, onListChange, onErrorChange]);

	const handleTextChange = useCallback(
		text => {
			if (error) {
				onErrorChange(null);
			}
			setText(text);
			onTextChange(!!text);
		},
		[error, onTextChange, onErrorChange]
	);

	const handleRemoveClick = useCallback(item => onListChange(list.filter(old => old !== item)), [list, onListChange]);

	const handleHighlightClear = useCallback(() => setHighlightKey(null), []);

	return (
		<div className="ListInput">
			<div className="ListInput__label">
				<Label htmlFor={id} text={label} />
				{notice && <div className="ListInput__notice">{notice}</div>}
			</div>
			<ComboBox
				{...props}
				id={id}
				autoSelect={!parseInput}
				text={text}
				error={!!error}
				getSuggestions={handleGetSuggestions}
				renderSuggestion={renderSuggestion}
				onChange={handleChange}
				onEnter={handleEnter}
				onTextChange={handleTextChange}
			/>
			<FieldError id={id + '-error'} text={error} />
			{list && list.length !== 0 ? (
				<ListEntries
					id={id + '-grid'}
					highlightKey={highlightKey}
					list={list}
					column={column}
					getItemKey={getItemKey}
					renderItem={renderItem}
					onRemoveClick={handleRemoveClick}
					onHighlightClear={handleHighlightClear}
				/>
			) : (
				<div className="ListInput__empty" id={id + '-empty'}>
					{emptyText}
				</div>
			)}
		</div>
	);
};

ListInput.propTypes = {
	id: PropTypes.string.isRequired,
	label: PropTypes.string,
	notice: PropTypes.string,
	placeholder: PropTypes.string,
	emptyText: PropTypes.string,
	list: PropTypes.array,
	error: PropTypes.string,
	column: PropTypes.bool,
	getSuggestions: PropTypes.func,
	getSuggestionText: PropTypes.func,
	getHighlightKey: PropTypes.func,
	parseInput: PropTypes.func,
	validateSuggestion: PropTypes.func,
	renderSuggestion: PropTypes.func,
	getItemKey: PropTypes.func,
	renderItem: PropTypes.func,
	onListChange: PropTypes.func,
	onTextChange: PropTypes.func,
	onErrorChange: PropTypes.func,
};

export default ListInput;
