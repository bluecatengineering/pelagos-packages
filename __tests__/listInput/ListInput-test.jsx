import React, {useState} from 'react';
import {shallow} from 'enzyme';

import ListInput from '../../src/listInput/ListInput';

jest.unmock('../../src/listInput/ListInput');

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useState: jest.fn(v => [v, jest.fn()]),
}));

describe('ListInput', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<ListInput
					id="test"
					label="Test"
					notice="Notice"
					placeholder="test placeholder"
					emptyText="Test empty"
					list={[{id: '0', name: 'test'}]}
					text="test text"
					error="test error"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when the notice is not set', () => {
			const wrapper = shallow(
				<ListInput
					id="test"
					label="Test"
					placeholder="test placeholder"
					emptyText="Test empty"
					list={[{id: '0', name: 'test'}]}
					text="test text"
					error="test error"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when the list is empty', () => {
			const wrapper = shallow(
				<ListInput
					id="test"
					label="Test"
					notice="Notice"
					placeholder="test placeholder"
					emptyText="Test empty"
					list={[]}
					text="test text"
					error="test error"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when error is not set', () => {
			const wrapper = shallow(
				<ListInput
					id="test"
					label="Test"
					notice="Notice"
					placeholder="test placeholder"
					emptyText="Test empty"
					list={[{id: '0', name: 'test'}]}
					text="test text"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls getSuggestions when the combo-box requests suggestions', () => {
			const suggestions = [{}];
			const list = [{}];
			const getSuggestions = jest.fn().mockReturnValue({suggestions});
			const wrapper = shallow(<ListInput id="test" list={list} getSuggestions={getSuggestions} />);
			expect(wrapper.find('ComboBox').prop('getSuggestions')('x')).toBe(suggestions);
			expect(getSuggestions.mock.calls).toEqual([['x', list]]);
		});

		it('calls getSuggestions and onErrorChange when the combo-box requests suggestions and the result is an error', () => {
			const error = 'Error';
			const list = [{}];
			const getSuggestions = jest.fn().mockReturnValue({error});
			const onErrorChange = jest.fn();
			const wrapper = shallow(
				<ListInput id="test" list={list} getSuggestions={getSuggestions} onErrorChange={onErrorChange} />
			);
			expect(wrapper.find('ComboBox').prop('getSuggestions')('x')).toEqual([]);
			expect(getSuggestions.mock.calls).toEqual([['x', list]]);
			expect(onErrorChange.mock.calls).toEqual([['Error']]);
		});

		it('sets highlightKey when the combo-box value changes and getHighlightKey returns an id', () => {
			const suggestion = {name: 'Test suggestion'};
			const getHighlightKey = jest.fn().mockReturnValue(1);
			const onTextChange = jest.fn();
			const onErrorChange = jest.fn();
			useState.mockReturnValueOnce([null, jest.fn()]);
			const wrapper = shallow(
				<ListInput
					id="test"
					getHighlightKey={getHighlightKey}
					onTextChange={onTextChange}
					onErrorChange={onErrorChange}
				/>
			);
			wrapper.find('ComboBox').prop('onChange')(suggestion);
			expect(getHighlightKey.mock.calls).toEqual([[suggestion]]);
			expect(onTextChange.mock.calls).toEqual([['']]);
			expect(useState.mock.results[0].value[1].mock.calls).toEqual([[1]]);
		});

		it('calls onListChange when the combo-box value changes and parseInput is set', () => {
			const suggestion = {name: 'Test suggestion'};
			const getSuggestionText = jest.fn().mockReturnValue('Text');
			const getHighlightKey = jest.fn();
			const onListChange = jest.fn();
			const onTextChange = jest.fn();
			const wrapper = shallow(
				<ListInput
					id="test"
					list={['Foo']}
					getSuggestionText={getSuggestionText}
					getHighlightKey={getHighlightKey}
					parseInput={jest.fn()}
					onListChange={onListChange}
					onTextChange={onTextChange}
				/>
			);
			wrapper.find('ComboBox').prop('onChange')(suggestion);
			expect(getHighlightKey.mock.calls).toEqual([[suggestion]]);
			expect(onTextChange.mock.calls).toEqual([['']]);
			expect(onListChange.mock.calls).toEqual([[['Text', 'Foo']]]);
		});

		it('calls onErrorChange when the combo-box value changes and validateSuggestion returns an error', () => {
			const suggestion = {name: 'Test suggestion'};
			const getSuggestionText = jest.fn().mockReturnValue('Text');
			const getHighlightKey = jest.fn();
			const validateSuggestion = jest.fn().mockReturnValue('Error');
			const onTextChange = jest.fn();
			const onErrorChange = jest.fn();
			const wrapper = shallow(
				<ListInput
					id="test"
					getSuggestionText={getSuggestionText}
					getHighlightKey={getHighlightKey}
					validateSuggestion={validateSuggestion}
					renderSuggestion={() => <div />}
					onTextChange={onTextChange}
					onErrorChange={onErrorChange}
				/>
			);
			wrapper.find('ComboBox').prop('onChange')(suggestion);
			expect(getHighlightKey.mock.calls).toEqual([[suggestion]]);
			expect(validateSuggestion.mock.calls).toEqual([[suggestion]]);
			expect(getSuggestionText.mock.calls).toEqual([[suggestion]]);
			expect(onTextChange.mock.calls).toEqual([['Text']]);
			expect(onErrorChange.mock.calls).toEqual([['Error']]);
		});

		it('calls onListChange when the combo-box value changes and validateSuggestion returns null', () => {
			const item = {name: 'Item'};
			const suggestion = {name: 'Test suggestion'};
			const getHighlightKey = jest.fn();
			const validateSuggestion = jest.fn();
			const onListChange = jest.fn();
			const onTextChange = jest.fn();
			const wrapper = shallow(
				<ListInput
					id="test"
					list={[item]}
					getHighlightKey={getHighlightKey}
					validateSuggestion={validateSuggestion}
					renderSuggestion={() => <div />}
					onListChange={onListChange}
					onTextChange={onTextChange}
				/>
			);
			wrapper.find('ComboBox').prop('onChange')(suggestion);
			expect(getHighlightKey.mock.calls).toEqual([[suggestion]]);
			expect(validateSuggestion.mock.calls).toEqual([[suggestion]]);
			expect(onTextChange.mock.calls).toEqual([['']]);
			expect(onListChange.mock.calls).toEqual([[[suggestion, item]]]);
		});

		it('calls onListChange with an empty list when enter is pressed and test is /clear', () => {
			const onListChange = jest.fn();
			const onTextChange = jest.fn();
			const wrapper = shallow(
				<ListInput id="test" text="/clear" onListChange={onListChange} onTextChange={onTextChange} />
			);
			wrapper.find('ComboBox').prop('onEnter')();
			expect(onTextChange.mock.calls).toEqual([['']]);
			expect(onListChange.mock.calls).toEqual([[[]]]);
		});

		it('calls onErrorChange when enter is pressed parseInput returns an error', () => {
			const list = [{}];
			const error = 'Error';
			const parseInput = jest.fn().mockReturnValue({error});
			const onErrorChange = jest.fn();
			const wrapper = shallow(
				<ListInput id="test" list={list} text="test" parseInput={parseInput} onErrorChange={onErrorChange} />
			);
			wrapper.find('ComboBox').prop('onEnter')();
			expect(parseInput.mock.calls).toEqual([['test', list]]);
			expect(onErrorChange.mock.calls).toEqual([[error]]);
		});

		it('calls onListChange with the parsed list when enter is pressed and parseInput returns a list', () => {
			const list = [{id: 0}];
			const entries = [{id: 1}];
			const parseInput = jest.fn().mockReturnValue({entries});
			const onListChange = jest.fn();
			const onTextChange = jest.fn();
			const wrapper = shallow(
				<ListInput
					id="test"
					list={list}
					text="test"
					parseInput={parseInput}
					onListChange={onListChange}
					onTextChange={onTextChange}
				/>
			);
			wrapper.find('ComboBox').prop('onEnter')();
			expect(parseInput.mock.calls).toEqual([['test', list]]);
			expect(onTextChange.mock.calls).toEqual([['']]);
			expect(onListChange.mock.calls).toEqual([[[...entries, ...list]]]);
		});

		it('ignores input when enter is pressed and parseInput is not set', () => {
			const onListChange = jest.fn();
			const onTextChange = jest.fn();
			const wrapper = shallow(<ListInput id="test" onListChange={onListChange} onTextChange={onTextChange} />);
			wrapper.find('ComboBox').prop('onEnter')();
		});

		it('calls onTextChange when the combo-box text changes', () => {
			const text = 'test';
			const onTextChange = jest.fn();
			const wrapper = shallow(<ListInput id="test" onTextChange={onTextChange} />);
			wrapper.find('ComboBox').prop('onTextChange')(text);
			expect(onTextChange.mock.calls).toEqual([[text]]);
		});

		it('calls onTextChange and onErrorChange when the combo-box text changes and error is set', () => {
			const text = 'test';
			const onTextChange = jest.fn();
			const onErrorChange = jest.fn();
			const wrapper = shallow(
				<ListInput id="test" error="Error" onTextChange={onTextChange} onErrorChange={onErrorChange} />
			);
			wrapper.find('ComboBox').prop('onTextChange')(text);
			expect(onTextChange.mock.calls).toEqual([[text]]);
			expect(onErrorChange.mock.calls).toEqual([[null]]);
		});

		it('calls onListChange when an item is removed', () => {
			const item0 = {id: 0};
			const item1 = {id: 1};
			const onListChange = jest.fn();
			const wrapper = shallow(<ListInput id="test" list={[item0, item1]} onListChange={onListChange} />);
			wrapper.find('ListEntries').prop('onRemoveClick')(item0);
			expect(onListChange.mock.calls).toEqual([[[item1]]]);
		});

		it('sets highlightKey to null when highlight is cleared', () => {
			const wrapper = shallow(<ListInput id="test" list={[{}]} />);
			wrapper.find('ListEntries').prop('onHighlightClear')();
			expect(useState.mock.results[0].value[1].mock.calls).toEqual([[null]]);
		});
	});
});
