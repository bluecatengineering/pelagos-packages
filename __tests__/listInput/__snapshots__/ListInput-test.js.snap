// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListInput rendering renders expected elements 1`] = `
<div
  className="ListInput"
>
  <LabelLine
    htmlFor="test"
    notice="Notice"
    optionalText="optional"
    showOptionalText={false}
    text="Test"
  />
  <ComboBox
    autoSelect={true}
    error={true}
    getSuggestions={[Function]}
    id="test"
    onChange={[Function]}
    onEnter={[Function]}
    onTextChange={[Function]}
    placeholder="test placeholder"
    text=""
  />
  <FieldError
    alignment="right"
    id="test-error"
    text="test error"
  />
  <ListEntries
    highlightKey={null}
    id="test-grid"
    list={
      Array [
        Object {
          "id": "0",
          "name": "test",
        },
      ]
    }
    onHighlightClear={[Function]}
    onRemoveClick={[Function]}
  />
</div>
`;

exports[`ListInput rendering renders expected elements when column is set 1`] = `
<div
  className="ListInput"
>
  <LabelLine
    htmlFor="test"
    notice="Notice"
    showOptionalText={false}
    text="Test"
  />
  <ComboBox
    autoSelect={true}
    error={true}
    getSuggestions={[Function]}
    id="test"
    onChange={[Function]}
    onEnter={[Function]}
    onTextChange={[Function]}
    placeholder="test placeholder"
    text=""
  />
  <FieldError
    alignment="right"
    id="test-error"
    text="test error"
  />
  <ListEntries
    column={true}
    highlightKey={null}
    id="test-grid"
    list={
      Array [
        Object {
          "id": "0",
          "name": "test",
        },
      ]
    }
    onHighlightClear={[Function]}
    onRemoveClick={[Function]}
  />
</div>
`;

exports[`ListInput rendering renders expected elements when error is not set 1`] = `
<div
  className="ListInput"
>
  <LabelLine
    htmlFor="test"
    notice="Notice"
    showOptionalText={false}
    text="Test"
  />
  <ComboBox
    autoSelect={true}
    error={false}
    getSuggestions={[Function]}
    id="test"
    onChange={[Function]}
    onEnter={[Function]}
    onTextChange={[Function]}
    placeholder="test placeholder"
    text=""
  />
  <FieldError
    alignment="right"
    id="test-error"
  />
  <ListEntries
    highlightKey={null}
    id="test-grid"
    list={
      Array [
        Object {
          "id": "0",
          "name": "test",
        },
      ]
    }
    onHighlightClear={[Function]}
    onRemoveClick={[Function]}
  />
</div>
`;

exports[`ListInput rendering renders expected elements when the list is empty 1`] = `
<div
  className="ListInput"
>
  <LabelLine
    htmlFor="test"
    notice="Notice"
    optionalText="optional"
    showOptionalText={true}
    text="Test"
  />
  <ComboBox
    autoSelect={true}
    error={true}
    getSuggestions={[Function]}
    id="test"
    onChange={[Function]}
    onEnter={[Function]}
    onTextChange={[Function]}
    placeholder="test placeholder"
    text=""
  />
  <FieldError
    alignment="right"
    id="test-error"
    text="test error"
  />
  <div
    className="ListInput__empty"
    id="test-empty"
  >
    Test empty
  </div>
</div>
`;
